File: YYReader/Services/HostRateLimiter.swift
import Foundation

actor HostRateLimiter {
    private var nextAvailable: [String: ContinuousClock.Instant] = [:]
    private let defaultMinimumDelay: Duration
    private var hostMinimumDelays: [String: Duration]

    init(
        defaultMinimumDelay: Duration = .seconds(1),
        hostMinimumDelays: [String: Duration] = [:]
    ) {
        self.defaultMinimumDelay = defaultMinimumDelay
        self.hostMinimumDelays = Dictionary(
            uniqueKeysWithValues: hostMinimumDelays.map { ($0.key.lowercased(), $0.value) }
        )
    }

    func setMinimumDelay(_ delay: Duration, forHost host: String) {
        hostMinimumDelays[host.lowercased()] = delay
    }

    func wait(for url: URL) async throws {
        let host = url.host?.lowercased() ?? ""
        let clock = ContinuousClock()
        let now = clock.now
        let delay = hostMinimumDelays[host] ?? defaultMinimumDelay
        let scheduled = max(now, nextAvailable[host] ?? now)
        nextAvailable[host] = scheduled.advanced(by: delay)

        if scheduled > now {
            try await clock.sleep(until: scheduled)
        }
    }
}

File: windows/YYReader.Windows.Core/Services/HostRateLimiter.cs
namespace YYReader.Windows.Core.Services;

public sealed class HostRateLimiter
{
    private readonly object _gate = new();
    private readonly TimeSpan _defaultMinimumDelay;
    private readonly Dictionary<string, TimeSpan> _hostMinimumDelays;
    private readonly Dictionary<string, DateTimeOffset> _nextAvailable = new(StringComparer.OrdinalIgnoreCase);

    public HostRateLimiter(
        TimeSpan? defaultMinimumDelay = null,
        IReadOnlyDictionary<string, TimeSpan>? hostMinimumDelays = null)
    {
        _defaultMinimumDelay = defaultMinimumDelay ?? TimeSpan.FromSeconds(1);
        _hostMinimumDelays = hostMinimumDelays is null
            ? new Dictionary<string, TimeSpan>(StringComparer.OrdinalIgnoreCase)
            : hostMinimumDelays.ToDictionary(pair => pair.Key, pair => pair.Value, StringComparer.OrdinalIgnoreCase);
    }

    public void SetMinimumDelay(string host, TimeSpan delay)
    {
        lock (_gate)
        {
            _hostMinimumDelays[host] = delay;
        }
    }

    public async Task WaitAsync(Uri url, CancellationToken cancellationToken = default)
    {
        var host = url.DnsSafeHost;
        DateTimeOffset scheduled;
        lock (_gate)
        {
            var now = DateTimeOffset.UtcNow;
            var delay = _hostMinimumDelays.GetValueOrDefault(host, _defaultMinimumDelay);
            scheduled = _nextAvailable.TryGetValue(host, out var current) && current > now ? current : now;
            _nextAvailable[host] = scheduled.Add(delay);
        }

        var wait = scheduled - DateTimeOffset.UtcNow;
        if (wait > TimeSpan.Zero)
        {
            await Task.Delay(wait, cancellationToken).ConfigureAwait(false);
        }
    }
}
